// react-async-component/commonjs/ssr/ClientProvider.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var ClientProvider = function (_React$Component) {
  _inherits(ClientProvider, _React$Component);

  function ClientProvider(props) {
    _classCallCheck(this, ClientProvider);

    var _this = _possibleConstructorReturn(this, (ClientProvider.__proto__ || Object.getPrototypeOf(ClientProvider)).call(this, props));

    _this.id = 0;
    _this.registry = {};
    return _this;
  }

  _createClass(ClientProvider, [{
    key: 'getChildContext',
    value: function getChildContext() {
      var _this2 = this;

      return {
        asyncComponents: {
          nextId: function nextId() {
            _this2.id += 1;
            return _this2.id;
          },
          registerComponent: function registerComponent(id, Component) {
            _this2.props.execContext.registerComponent(id, Component);
          },
          getComponent: function getComponent(id) {
            return _this2.props.execContext.getComponent(id);
          }
        }
      };
    }
  }, {
    key: 'render',
    value: function render() {
      return _react2.default.Children.only(this.props.children);
    }
  }]);

  return ClientProvider;
}(_react2.default.Component);

ClientProvider.propTypes = {
  children: _react2.default.PropTypes.node.isRequired,
  execContext: _react2.default.PropTypes.shape({
    registerComponent: _react2.default.PropTypes.func.isRequired,
    getComponent: _react2.default.PropTypes.func.isRequired
  }).isRequired
};

ClientProvider.childContextTypes = {
  asyncComponents: _react2.default.PropTypes.shape({
    nextId: _react2.default.PropTypes.func.isRequired,
    registerComponent: _react2.default.PropTypes.func.isRequired,
    getComponent: _react2.default.PropTypes.func.isRequired
  }).isRequired
};

exports.default = ClientProvider;
